src/libs/actions/OnyxUpdateManager.ts

```diff
@@ -66,32 +70,114 @@ export default () => {
+
+                // We only want to apply deferred updates that are newer than the last update that was applied to the client.
+                // At this point, the missing updates from "GetMissingOnyxUpdates" have been applied already, so we can safely filter out.
+                const pendingDeferredUpdates = Object.fromEntries(
+                    Object.entries(deferredUpdates).filter(([lastUpdateID]) => {
+                        // Same as above, it should not be possible for lastUpdateIDAppliedToClient to be null.
+                        // Still, if so we want to keep the deferred update in the list
+                        if (!lastUpdateIDAppliedToClient) {
+                            return true;
+                        }
+                        return (Number(lastUpdateID) ?? 0) > lastUpdateIDAppliedToClient;
+                    }),
+                );
+                const pendingDeferredUpdateValues = Object.values(pendingDeferredUpdates);
+
+                function unpauseQueueAndReset() {
                     console.debug('[OnyxUpdateManager] Done applying all updates');
                     Onyx.set(ONYXKEYS.ONYX_UPDATES_FROM_SERVER, null);
                     SequentialQueue.unpause();
+                }
+
+                // If there are no remaining deferred updates after filtering out outdated ones,
+                //  we can just unpause the queue and return
+                if (pendingDeferredUpdateValues.length === 0) {
+                    unpauseQueueAndReset();
+                    return;
+                }
+
+                let lastValidDeferredUpdateID = 0;
+                // In order for the deferred updates to be applied correctly in order,
+                // we need to check if there are any gaps deferred updates.
+                function areDeferredUpdatesChained(): boolean {
+                    let isFirst = true;
+                    for (const update of pendingDeferredUpdateValues) {
+                        // If it's the first one, we need to skip it since we won't have the previousUpdateID
+                        // if the previousUpdateID key exists, then we can just move on since there's no gap
+                        if (isFirst || pendingDeferredUpdates[Number(update.previousUpdateID)]) {
+                            lastValidDeferredUpdateID = Number(update.lastUpdateID);
+                            isFirst = false;
+                            // eslint-disable-next-line no-continue
+                            continue;
+                        }
+
+                        return false;
+                    }
+
+                    return true;
+                }
+
+                //  If we detect a gap in the deferred updates, re-fetch the missing updates.
+                if (!areDeferredUpdatesChained()) {
+                    canUnpauseQueuePromise = App.getMissingOnyxUpdates(lastUpdateIDAppliedToClient, lastValidDeferredUpdateID).finally(applyDeferredUpdates);
```

Comment:
```suggestion
                    canUnpauseQueuePromise = App.getMissingOnyxUpdates(lastUpdateIDAppliedToClient, lastValidDeferredUpdateID);
```

hmmm this is not exactly what we need to do. If you do getMissingOnyxUpdates, it will apply the updates it will receive directly to onyx.

So let's say:

- Your current state is lastUpdateID 10
- You receive through pusher update A with previousUpdateID 15, lastUpdateID 18
- You then store that update and start a Request 1, getMissingOnyxMessages, from 10 to 15
- While you're executing Request 1, you receive update B with previousUpdateID 20 and lastUpdateID 22, which is stored in the deferredUpdates
- When Request 1 finishes, updates from 10 to 15 will be applied directly into onyx, and the local lastUpdateID is now 15
- You'll run this code, and see that there's a gap between update A (lastUpdateID 18) and update B (previousUpdateID 20)
- Because that gap exists, you'll do another getMissingOnyxUpdates, Request 2, from 18 to 20
- When request 2 finishes, it will return previousUpdateID 18 and lastUpdateID 20, and try to automatically apply that updates received through HTTPs before applying anything from the deferred queue. It will then notice that the local lastUpdateID is 15, which doesn't match the received previousUpdateID 18. It will query another GetMissingOnyxMessages from 15 to 18 (but this already exists in the deferredUpdates

So this will still trigger more getMissingOnyxUpdates, even if we have those updates ready to be applied.

Reply:
You're right. I updated the logic. Can you give this another review? :)